Seminar2

KWY0218 opened the pull request:
세미나 2주차 과제입니다.

[ReadMe](https://github.com/SAMSUNG-ZZANG/Galaxy-Tab-Wonyong/blob/seminar2/README.md)

KWY0218 committed:
1st commit Seminar2 assignment

KWY0218 committed:
[Refactoring] edit adapter

KWY0218 committed:
Update README.md

KWY0218 committed:
[Refactoring] edit submit list

KWY0218 committed:
Update README.md

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/data/source/local/UserLocalDatSource.kt:
```diff
@@ -1,6 +1,6 @@
 package com.example.sopt_seminar.data.source.local
 
-import com.example.sopt_seminar.data.model.User
+import com.example.sopt_seminar.domain.model.User
```
데이터 소스를 도메인 패키지에 넣는 이유가 무엇인지 알 수 있을까요? 도메인의 경우 데이터로직과 비니지스 로직을 분리하기 위해서(그래서 레포지토리나 UseCase등이 들어감.) 정의되는 레이어라고 생각하는데, 데이터 소스의 경우 그런 역할을 하는 친구는 아니고 말 그대로, 데이터소스(=데이터의 출발지)라고 생각하는데, 데이터 레이어에 넣는게 아키텍쳐 관점에서 좋을 것 같지않나요??

KWY0218 replied:
`SignUpViewModel` 에서 `User` 를 사용하는데, ui layer에서 data layer에 있는 model을 사용하는 것과
도메인에서는 데이터 레이어에 대해 몰라야 하는데 domain layer에 있는 UseCase에서 `User` 를 사용하므로 도메인으로 옮겼습니다.

생각해보니까 `SignUpViewModel` 에서  `setUserUseCase(User(nameText, userId, userPassword))` 이런 식으로 보냈는데, 그냥 nameText, userId, userPassword 값만 넘겨주고, usecase에서 `User`를 넣는 것으로 수정했습니다.

data layer에도 `UserEntity` model을 만들어서 domain 에서 유저 정보가 필요할 땐  매핑 하는 것으로 수정했습니다.

mdb1217 replied:
제가 정신이 나가서 데이터소스를 도메인에 옮겼다고 생각하고 여기다 리뷰를 단것같네요.. 이부분은 무시해주셔도 될것같습니다..

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/ui/FollowerFragment.kt:
```diff
@@ -0,0 +1,90 @@
+
+        with(binding) {
+            followerRecyclerView.adapter = adapter
+            followerRecyclerView.layoutManager = LinearLayoutManager(context)
```
레이아웃 매니저의 경우 xml 상에서 선언할 수 있으니까 거기서 작성해도 괜찮을 것 같아요..!

KWY0218 replied:
요새 bindingAdapter 공부 중인데, 후에 적용해볼 때 수정하겠습니다!!

mdb1217 commented on README.md:
```diff
@@ -1,218 +1,269 @@
+
+# Repo, Follower Adapter
+- ListAdapter로 구현
+- ListAdapter는 Recyclerview Adapter가 기반인 클래스이고 아이템들 사이의 차이점을 백스라운드에서 계산을 한다.
```
정확히는 ListAdapter내부에 `AsyncDiffer`가 내장되어있어서 백그라운드 계산이 가능한 것입니다. 리사이클러뷰 + 디프유틸 + AsyncDiffer를 같이 사용하면 동일 성능을 낼 수 있어요! 다만 구현할 객체들이 많아지는 게 단점일뿐..ㅎ

KWY0218 replied:
제가 너무 대강 공부한 거 같네요..ㅠㅠ 감사합니다..!

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/ui/FollowerFragment.kt:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    companion object {
+        private val testList: MutableList<Follower> = mutableListOf(
```
```suggestion
        private val testList = mutableListOf(
```
코틀린에서는 타입추론 기능이 있기때문에 이렇게만 표현하셔도 됩니다!

KWY0218 replied:
수정했습니다!!

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/ui/HomeFragment.kt:
```diff
@@ -1,28 +1,39 @@
+            .commit()
+
+        with(binding) {
+            homeTab.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
+                override fun onTabSelected(tab: TabLayout.Tab) {
+                    val transaction = childFragmentManager.beginTransaction()
+                    when (tab.position) {
+                        FOLLOWER_TURN -> transaction.replace(R.id.home_fragment_cv, followerFragment)
+                        REPOSITORY_TURN -> transaction.replace(R.id.home_fragment_cv, repoFragment)
+                    }
+                    transaction.commit()
+                }
+                override fun onTabUnselected(tab: TabLayout.Tab) = Unit
+                override fun onTabReselected(tab: TabLayout.Tab) = Unit
+            })
```
혹시 tabLayout을 뷰페이저가 아닌 프래그먼트 콘테이너뷰에 연동시킨 이유가 있을까요?? 탭레이아웃과 뷰페이저를 연동하면 탭레이아웃 메디에이터를 사용가능하니까요!

KWY0218 replied:
과제에 ContainerView 사용이라고 나와있어서 viewPager는 적용을 안 했었고, 버튼 2개 보다 텝레이아웃으로 해보고 싶어서 텝레이아웃만 사용했었습니다 😅

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/ui/RepoFragment.kt:
```diff
@@ -0,0 +1,89 @@
+                viewHolder: RecyclerView.ViewHolder,
+                target: RecyclerView.ViewHolder
+            ): Boolean {
+                val startPosition = viewHolder.adapterPosition
```
`viewHolder.adapterPosition`은 deprecated 되었어요! 아래링크 참고바람! 

https://developer.android.com/reference/androidx/recyclerview/widget/RecyclerView.ViewHolder

KWY0218 replied:
![image](https://user-images.githubusercontent.com/82709044/165754487-d834b6fc-c9b3-4123-8d39-97510dd6e25e.png)

문서에서 deprecated 된 것을 확인했는데, 사진처럼 아직 밑줄이 안 그어져 있습니다..! 그냥 업데이트가 아직 안된 것일까요???

현실 부정하지 말고 리팩토링 하면 되나요???

mdb1217 replied:
<img width="792" alt="스크린샷 2022-04-29 오전 1 27 20" src="https://user-images.githubusercontent.com/70698151/165800241-a3e38064-86e1-41e0-a83b-027c97914e84.png">

:(

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/ui/adapter/FollowerAdapter.kt:
```diff
@@ -0,0 +1,77 @@
+class FollowerAdapter: ListAdapter<Follower, FollowerAdapter.FollowerViewHolder>(FOLLOWER_COMPARATOR){
+
+    private lateinit var itemClick: OnItemClickListener
+    interface OnItemClickListener{
+        fun onItemClick(position: Int)
+    }
```
인터페이스 말고 고차함수를 활용해서 좀 더 간결하게 표현할 수도 있어용..!

KWY0218 replied:
와!!! 최고!!!!

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/util/BaseFragment.kt:
```diff
@@ -0,0 +1,37 @@
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        binding.lifecycleOwner = this
```
프래그먼트의 lifecycleOnwer를 this로 지정하면 `onCreate-onDestroy` 범위 까지 라이프사이클 오너가 지정되므로

```suggestion
        binding.lifecycleOwner = viewLifeCycleOwner
```

로 지정하는게 좋습니다. 프래그먼트 생명주기 문제로 이것 또한 메모리 릭이 발생할 수 있으니 참고해주세요!

링크 : https://pluu.github.io/blog/android/2020/01/25/android-fragment-lifecycle/

KWY0218 replied:
정독했습니다!! 감사합니다!!!

mdb1217 commented on app/src/main/java/com/example/sopt_seminar/ui/adapter/FollowerAdapter.kt:
```diff
@@ -0,0 +1,77 @@
+        }
+
+        fun bind(_name: String, _des: String) {
+            name.text = _name
+            des.text = _des
```
데이터 바인딩의 특성을 활용해서 레이아웃에서 변수 선언해서 한 번에 집어넣는 식으로 하는 것도 깔끔할듯요!

mdb1217 commented on app/src/main/res/layout/home_fragment.xml:
```diff
@@ -56,24 +55,44 @@
+            <androidx.fragment.app.FragmentContainerView
+                android:id="@+id/home_fragment_cv"
+                android:layout_width="match_parent"
+                android:layout_height="450dp"
```
0dp로 걸고 탭 레이아웃에 Top 제약 걸고 Bottom 부모로 제약 걸면 굳이 고정값 안넣어도 괜찮을 것 같습니다!

KWY0218 replied:
처음에 0dp 했다가 스크롤뷰 때문에 리사이클러뷰로 스크롤하는 것이 아니라 스크롤뷰 때문에 스크롤되는 느낌이 들어서 높이를 고정했었습니다..!

KWY0218 committed:
[style] delete mutable list type

KWY0218 committed:
[refactor] edit basefragment lifecycleowner

KWY0218 committed:
[refactor] edit listadapter click event

interface 에서 고차함수로 변경

KWY0218 committed:
[refactor] add userEntity in data package, edit setuserusecase

repo와 follower는 로컬 저장소를 사용할 때 추가할 예정입니다.

KWY0218 committed:
[REFACTOR] Edit adapter

KWY0218 committed:
[REFACTOR] Delete tab layout

KWY0218 committed:
[REFACTOR] Edit adapter

KWY0218 committed:
[REFACTOR] recyclerview bindingAdapter

- onViewCreated 깔끔해지고 싶어서 bindAdapter 사용했다
- 이게 잘 활용한건지 남용인지 구분이 안간다
- adapterPosition 은 다음에...ㅠㅠ
